packages/protocol/contracts/ACE/validators/adjustSupply/AdjustSupply.sol

```diff
@@ -0,0 +1,384 @@
+                mstore(0x80, calldataload(0x44))
+                mstore(0xa0, calldataload(0x64))
+                let notes := add(0x104, calldataload(0x144)) // get the length of notes
+                let m := 1
+                let n := calldataload(notes)
+                let gen_order := 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
+                let challenge := mod(calldataload(0x124), gen_order)
+
+
+                // validate m <= n
+                if gt(m, n) { mstore(0x00, 404) revert(0x00, 0x20) }
+                
+                let kn := calldataload(sub(add(notes, mul(calldataload(notes), 0xc0)), 0xa0))
+
+                // add sender final hash table
+                mstore(0x2a0, calldataload(0x24))
+
+                kn := mulmod(sub(gen_order, kn), challenge, gen_order) // we actually want c*k_{public}
+
+
+                hashCommitments(notes, n)  // notes = length of proof data array, n = number of notes
+                let b := add(0x2c0, mul(n, 0x80))
+
+
+                // Iterate over every note and calculate the blinding factor B_i = \gamma_i^{kBar}h^{aBar}\sigma_i^{-c}.
+                // We use the AZTEC protocol pairing optimization to reduce the number of pairing comparisons to 1,
+                //  which adds some minor alterations
+                for { let i := 0 } lt(i, n) { i := add(i, 0x01) } {
+
+                // Get the calldata index of this note
+                let noteIndex := add(add(notes, 0x20), mul(i, 0xc0))
+
+                // Define variables k, a and c.
+                // If i <= m then
+                //   k = kBar_i
+                //   a = aBar_i
+                //   c = challenge
+                // If i > m then we add a modification for the pairing optimization
+                //   k = kBar_i * x_i
+                //   a = aBar_i * x_i
+                //   c = challenge * x_i
+                // Set j = i - (m + 1).
+                // x_0 = 1
+                // x_1 = keccak256(input string)
+                // all other x_{j} = keccak256(x_{j-1})
+                // The reason for doing this is that the point  \sigma_i^{-cx_j} can be re-used in the pairing check
+                // Instead of validating e(\gamma_i, t_2) == e(\sigma_i, g_2) for all i = [m+1,\ldots,n]
+                // We validate e(\Pi_{i=m+1}^{n}\gamma_i^{-cx_j}, t_2) == e(\Pi_{i=m+1}^{n}\sigma_i^{cx_j}, g_2).
+                // x_j is a pseudorandom variable whose entropy source is the input string, allowing for
+                // a sum of commitment points to be evaluated in one pairing comparison
+                let k
+                let a := calldataload(add(noteIndex, 0x20))
+                let c := challenge
+                
+                switch eq(add(i, 0x01), n)
+                    case 1 {
+                        k := kn
+
+                        // if all notes are input notes, invert k
+                        if eq(m, n) {
```

Comment:
we can remove this `if` condition as we should be throwing an error if this occurs

Reply:
Good point, done